crates/typst-syntax/src/ast.rs

```diff
@@ -553,83 +551,55 @@ node! {
-            if s.eat_if(is_id_start) {
-                s.eat_while(is_id_continue);
-            }
-            text = s.after();
-            text = text.strip_prefix(' ').unwrap_or(text);
-            if text.trim_end().ends_with('`') {
-                text = text.strip_suffix(' ').unwrap_or(text);
-            }
-        }
-
-        // Split into lines.
-        let mut lines = split_newlines(text);
-
-        if blocky {
-            let dedent = lines
-                .iter()
-                .skip(1)
-                .filter(|line| !line.chars().all(char::is_whitespace))
-                // The line with the closing ``` is always taken into account
-                .chain(lines.last())
-                .map(|line| line.chars().take_while(|c| c.is_whitespace()).count())
-                .min()
-                .unwrap_or(0);
-
-            // Dedent based on column, but not for the first line.
-            for line in lines.iter_mut().skip(1) {
-                let offset = line.chars().take(dedent).map(char::len_utf8).sum();
-                *line = &line[offset..];
-            }
-
-            let is_whitespace = |line: &&str| line.chars().all(char::is_whitespace);
-
-            // Trims a sequence of whitespace followed by a newline at the start.
-            if lines.first().map_or(false, is_whitespace) {
-                lines.remove(0);
-            }
-
-            // Trims a newline followed by a sequence of whitespace at the end.
-            if lines.last().map_or(false, is_whitespace) {
-                lines.pop();
-            }
-        }
-
-        lines.join("\n").into()
+    /// The lines in the raw block.
+    pub fn lines(self) -> impl Iterator<Item = RawLine<'a>> {
+        self.0.children().filter_map(RawLine::from_untyped)
     }
 
     /// An optional identifier specifying the language to syntax-highlight in.
     pub fn lang(self) -> Option<&'a str> {
-        let text = self.0.text();
+        let delim: RawDelim = self.0.cast_first_match().unwrap();
 
         // Only blocky literals are supposed to contain a language.
-        if !text.starts_with("```") {
+        if delim.0.text().len() < 3 {
             return Option::None;
         }
 
-        let inner = text.trim_start_matches('`');
-        let mut s = Scanner::new(inner);
-        s.eat_if(is_id_start).then(|| {
-            s.eat_while(is_id_continue);
-            s.before()
-        })
+        let lang: Option<RawLang> = self.0.cast_first_match();
+        lang.map(|lang| lang.0.text().as_str())
     }
 
     /// Whether the raw text should be displayed in a separate block.
     pub fn block(self) -> bool {
-        let text = self.0.text();
-        text.starts_with("```") && text.chars().any(is_newline)
+        let delim: RawDelim = self.0.cast_first_match().unwrap();
+
+        let mut newlines = self.0.children().filter(|e| {
+            matches!(e.kind(), SyntaxKind::RawTrimmed) && e.text().chars().any(is_newline)
```

Comment:
does this behave the same as before? will the newline always be part of a RawTrimmed?

Reply:
See https://github.com/typst/typst/pull/3257#issuecomment-1912081856.

> When lexing and parsing content of a raw block, all effective content are stored in RawLines and rest of spaces are stored in RawTrimmed.

> From the code, whenever it meets a newline character, a RawLine and a RawTrimmed are created soonly. So, obviously, we won't have newline in the RawLine. If some RawLine indeed have one, there must be some bug in the lexer.